enhancements/cluster-logging/fluentd-tweaks.md

```diff
@@ -0,0 +1,166 @@
+---
+title: cluster-logging-log-forwarding
+authors:
+  - "@alanconway"
+reviewers:
+approvers:
+creation-date: 2020-05-27
+status: provisional
+see-also:[]
+replaces:[]
+superseded-by:[]
+---
+
+# Fluentd Peformance Tuning
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+### Naming of "tweaks"
+
+I've called these settings 'tweaks' (like the Gnome Tweaks application) to
+identify them clearly as things that most users don't need to set.  If there are
+objections to that, possible alternate names: 'knobs', 'hints', 'advanced',
+'expert', 'tuning' ... other?
+
+### Where to put the tweaks
+
+Below I describe tweaks as part of the Collector API, acting as defaults for all outputs. This is the way we current use environment variables.
+However while writing this I came to the conclusion that they may fit better as part of the ClusterLogForwarder.Output.
```

Comment:
I'm not sure I agree because LF is generally an abstraction over collection and writing to a destination.  Tuning is collector specific and may not apply if swap the collector.   I can see where we may tune individual outputs but what does it mean if the collector become ryslog (not likely); are thread and buffer tuning still applicable?  What do we say to user's when we swap the collector out from them and the settings no longer apply?  Maybe the question still stands regardless

Reply:
Agree 100% that this is specific to how we now use fluentd. If we change things it won't apply anymore, whether or not we introduce new tuning for the future thing. There's a note in the proposal that we don't give the same forward-compat guarantees for tuning as other API.

However, what I realized is we are tuning "sender" behavior. This is actually not "collector" behaviour - it just so happens that right now we use fluentd for both.

Suppose we had a fluentbit collector sending to an intermediary (loki, AMQP, kafka whatever) which in turn sends to Forwarder.Output.  Now if we want "sender" tuning it will be for the intermediary, nothing to do with the collector. So logically it seems like it belongs on the forwader.Output. It's also something that could be set per-Output rather than as a default.
 
The main problem I see with that is avoiding confusion between type-specific "destination" config (e.g. syslog.facility)  and "sender tuning"  (e.g. fluentd.chunk_size_limit) on the same output. I think we can design the API to separate them clearly, and it is an "advanced" feature so maybe that's not a big deal.

@jcantrill WDYT?

Reply:
Maybe I missed that but I would also vote against putting it into the log forwarder API and leave it where it is under the fluentd field in the cluster logging API. For me this makes much more sense and we consolidate all "collector" specific tunings into a single field and in case we change the collector, it's easier to spot things that we will not support anymore.

Individual collectors can have different semantics for certain tuning parameters as well as some that literally doesn't exist since it handles things differently to our current collector. So let's not spread any configuration and define them clearly as "fluentd specific" under a single configuration field.

The documentation should help in what cases what tuning parameter will help.

Reply:
3rd option: I think the right place is the ClusterLogging API (probably I created this confusion by proposing the Collector.) We're saying "Here are some global tuning hacks for the *current fluentd-based implementation* of logging." We're not implying such hacks are specifically related to collection or forwarding, they could be related to anything. And we're not dirtying the dedicated Collector/Forwarder APIs with concerns that don't belong there, or with temporary junk that should be replaced someday by clean extensions to the Collector/Forwarder APIs.

Making this change, shout if any objections!